### PR TITLE
fix example usage of wait on a record of futures

### PR DESCRIPTION
**Pull request by AbdeMohlbi**
Fixes <https://github.com/dart-lang/site-www/issues/6047>

---

- [x] I’ve reviewed the contributor guide and applied the relevant portions to this PR.
- [x] This PR doesn't contain automatically generated corrections or text (Grammarly, LLMs, and similar).
- [x] This PR follows the [Google Developer Documentation Style Guidelines](https://developers.google.com/style) — for example, it doesn't use _i.e._ or _e.g._, and it avoids _I_ and _we_ (first person).
- [x] This PR uses [semantic line breaks](https://github.com/dart-lang/site-shared/blob/main/doc/writing-for-dart-and-flutter-websites.md#semantic-line-breaks) of 80 characters or fewer.

<details>
  <summary>Contribution guidelines:</summary><br>

  - See our [contributor guide](https://github.com/dart-lang/site-www/blob/main/CONTRIBUTING.md) for general expectations for PRs.
  - Larger or significant changes should be discussed in an issue before creating a PR.
  - Code changes should generally follow the [Dart style guide](https://dart.dev/effective-dart) and use `dart format`.
  - Updates to [code excerpts](https://github.com/dart-lang/site-shared/blob/main/doc/code-excerpts.md) indicated by `<?code-excerpt` need to be updated in their source `.dart` file as well.
</details>

### CODE REVIEW

File: src/content/libraries/dart-async.md [hunk 1]
```diff
@@ -258,20 +258,22 @@ void main() async {
...
   Future<bool> errorResult() async =>  ...
-
+  int deleteInt;
+  String copyString;
+  bool errorBool;
   try {    
     // Wait for each future in a record, returns a record of futures:
-    (int, String, bool) result = await (delete(), copy(), errorResult()).wait;
-  
+    (deleteInt, copyString ,errorBool) = await (delete(), copy(), errorResult()).wait;
```
Comment (nate-thegrate) [inline src/content/libraries/dart-async.md:266]:
```suggestion
    (deleteInt, copyString, errorBool) = await (delete(), copy(), errorResult()).wait;
```

File: src/content/libraries/dart-async.md [hunk 2]
```diff
@@ -258,20 +258,22 @@ void main() async {
...
+
   } on ParallelWaitError<(int?, String?, bool?),
       (AsyncError?, AsyncError?, AsyncError?)> catch (e) {
     // ...
+    rethrow; // if catch ends normally then you will need to initialize deleteInt, copyString ,errorBool.
```
Comment (nate-thegrate) [inline src/content/libraries/dart-async.md:271]:
```suggestion
    rethrow; // if catch ends normally then you will need to initialize deleteInt, copyString, errorBool.
```

File: src/content/libraries/dart-async.md [hunk 3]
```diff
@@ -258,20 +258,22 @@ void main() async {
   Future<int> delete() async =>  ...
   Future<String> copy() async =>  ...
   Future<bool> errorResult() async =>  ...
-
+  int deleteInt;
+  String copyString;
+  bool errorBool;
```
Comment (nate-thegrate) [inline src/content/libraries/dart-async.md:263]:
```suggestion

  int deleteInt;
  String copyString;
  bool errorBool;

```

File: src/content/libraries/dart-async.md [hunk 4]
```diff
@@ -258,20 +258,22 @@ void main() async {
...
       (AsyncError?, AsyncError?, AsyncError?)> catch (e) {
     // ...
+    rethrow; // if catch ends normally then you will need to initialize deleteInt, copyString ,errorBool.
     }
 
   // Do something with the results:
-  var deleteInt  = result.$1;
-  var copyString = result.$2;
-  var errorBool  = result.$3;
+    print("$deleteInt $copyString $errorBool");
```
Comment (nate-thegrate) [inline src/content/libraries/dart-async.md:275]:
```suggestion
  print("$deleteInt $copyString $errorBool");
```

**Review by nate-thegrate — COMMENTED** [review]
I don't think there's anything wrong with the original example.
- The original shows how to use the `.wait` getter for a record of futures, and also what its relevant exception looks like.
- The version in this PR shows how to initialize the values even if there's an error.

<br>

Perhaps this example could show how to initialize the values, instead of adding a comment about it.

```dart
void main() async {
  Future<int> delete() async => ...
  Future<String> copy() async => ...
  Future<bool> errorResult() async => ...

  int deleteInt;
  String copyString;
  bool errorBool;
  
  try {
    (deleteInt, copyString, errorBool) = await (delete(), copy(), errorResult()).wait;
  } on ParallelWaitError<(int?, String?, bool?),
      (AsyncError?, AsyncError?, AsyncError?)> catch (e) {
    deleteInt = e.values.$1 ?? 0;
    copyString = e.values.$2 ?? '';
    errorBool = e.values.$3 ?? true;

    print(e.errors);
  }

  // ...
}
```

Then again, once you're set up to work with the appropriate `ParallelWaitError` object, perhaps the rest is self-explanatory (i.e. easy to achieve just by looking at autofill suggestions).

Overall, I think I have a slight preference toward leaving the example the way it is, but would be interested to hear other viewpoints.

**Comment by AbdeMohlbi** [comment]
> I don't think there's anything wrong with the original example.
> - The original shows how to use the `.wait` getter for a record of futures, and also what its relevant exception looks like.
> - The version in this PR discusses initializing the values even if there's an error.
> 
> <br>
> 
> Perhaps this example could show how to initialize the values, instead of adding a comment about it.
> 
> ```dart
> void main() async {
>   Future<int> delete() async => ...
>   Future<String> copy() async => ...
>   Future<bool> errorResult() async => ...
> 
>   int deleteInt;
>   String copyString;
>   bool errorBool;
>   
>   try {
>     (deleteInt, copyString, errorBool) = await (delete(), copy(), errorResult()).wait;
>   } on ParallelWaitError<(int?, String?, bool?),
>       (AsyncError?, AsyncError?, AsyncError?)> catch (e) {
>     deleteInt = e.values.$1 ?? 0;
>     copyString = e.values.$2 ?? '';
>     errorBool = e.values.$3 ?? true;
> 
>     print(e.errors);
>   }
> 
>   // ...
> }
> ```
> 
> Then again, once you're set up to work with the appropriate `ParallelWaitError` object, perhaps the rest is self-explanatory (i.e. easy to achieve just by looking at autofill suggestions).
> 
> Overall, I think I have a slight preference toward leaving the example the way it is, or ideally, to just add a single `rethrow;` to the catch block. But I'd be interested to hear other viewpoints.

I did that because in the original one the result wasn't accessible out of the catch block

**Comment by nate-thegrate** [comment]
Fair. But the PR could be as simple as

> just add a single `rethrow;` to the catch block

**Comment by nate-thegrate** [comment]
The assumption is that the reader has an understanding of how try/catch works and just needs guidance pertaining to how Dart's `.wait` getter works for a record of futures.

The amount of detail to put into a snippet like this one lacks a universally "correct" answer—some might find the extra details helpful, whereas others (myself included) much prefer being able to read something concisely-written and then build off of it in our own projects.

**Comment by AbdeMohlbi** [comment]
> The amount of detail to put into a snippet like this one lacks a universally "correct" answer—some might find the extra details helpful, whereas others (myself included) much prefer being able to read something concisely-written and then build off of it in our own projects.

Based

**Comment by AbdeMohlbi** [comment]
> Fair. But the PR could be as simple as
> 
> > just add a single `rethrow;` to the catch block
Hmmm i'll give it a try later then

**Comment by AbdeMohlbi** [comment]
@nate-thegrate i guess this is better what do u think ?  
```dart
await (delete(), copy(), errorResult())
      .wait
      .then((result) => print(result))
      .onError<
          ParallelWaitError<(int?, String?, bool?),
              (AsyncError?, AsyncError?, AsyncError?)>>((error, stackTrace) {
    throw (error.errors,stackTrace);
  });
```

**Comment by nate-thegrate** [comment]
> ```dart
> await (delete(), copy(), errorResult())
>       .wait
>       .then((result) => print(result))
>       .onError<
>           ParallelWaitError<(int?, String?, bool?),
>               (AsyncError?, AsyncError?, AsyncError?)>>((error, stackTrace) {
>     throw (error.errors,stackTrace);
>   });
> ```

Probably not, sorry, since this goes against an [Effective Dart guideline](https://dart.dev/effective-dart/usage#prefer-asyncawait-over-using-raw-futures).

<br>

My preference would be for the example to match [the above snippet](https://github.com/dart-lang/site-www/pull/6048#pullrequestreview-2248460566), or to add a simple "rethrow":

```diff
  try {    
    // Wait for each future in a record, returns a record of futures:
    (int, String, bool) result = await (delete(), copy(), errorResult()).wait;
  
  } on ParallelWaitError<(int?, String?, bool?),
      (AsyncError?, AsyncError?, AsyncError?)> catch (e) {
    // ...
+   rethrow;

-   }
+ } // might as well fix indentation too
```

**Comment by AbdeMohlbi** [comment]
> > ```dart
> > await (delete(), copy(), errorResult())
> >       .wait
> >       .then((result) => print(result))
> >       .onError<
> >           ParallelWaitError<(int?, String?, bool?),
> >               (AsyncError?, AsyncError?, AsyncError?)>>((error, stackTrace) {
> >     throw (error.errors,stackTrace);
> >   });
> > ```
> 
> Probably not, sorry, since this goes against an [Effective Dart guideline](https://dart.dev/effective-dart/usage#prefer-asyncawait-over-using-raw-futures).
> 
> <br>
> 
> My preference would be for the example to match [the above snippet](https://github.com/dart-lang/site-www/pull/6048#pullrequestreview-2248460566), or to add a simple "rethrow":
> 
> ```diff
>   try {    
>     // Wait for each future in a record, returns a record of futures:
>     (int, String, bool) result = await (delete(), copy(), errorResult()).wait;
>   
>   } on ParallelWaitError<(int?, String?, bool?),
>       (AsyncError?, AsyncError?, AsyncError?)> catch (e) {
>     // ...
> +   rethrow;
> 
> -   }
> + } // might as well fix indentation too
> ```
Hmm i'll draft a new pr then and close this , is it ok to ping u when this is done to review ?

**Comment by nate-thegrate** [comment]
> Hmm i'll draft a new pr then and close this , is it ok to ping u when this is done to review ?

Yeah, sounds good!

**Pull request closed without merging**